File: others/100-200/115. 不同的子序列.cpp
//
// Created by 123 on 25-5-24.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;
constexpr int DIR[4][2] =  {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

class Solution {
public:
  int numDistinct(string s, string t) {
    int n = s.size(), m = t.size();
    if (n < m) {
      return 0;
    }

    // 和答案无关的转移可能会溢出，从而报错
    // 为了避免报错，使用 unsigned
    vector f(n + 1, vector<unsigned>(m + 1));
    f[0][0] = 1;
    for (int i = 0; i < n; i++) {
      f[i + 1][0] = 1;
      for (int j = max(m - n + i, 0); j < min(i + 1, m); j++) {
        f[i + 1][j + 1] = f[i][j + 1];
        if (s[i] == t[j]) {
          f[i + 1][j + 1] += f[i][j];
        }
      }
    }
    return f[n][m];
  }
};

File: others/100-200/115. 不同的子序列.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

MOD = int(1e9 + 7)
DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]


class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n, m = len(s), len(t)
        if n < m:
            return 0

        f = [1] + [0] * m
        for i, x in enumerate(s):
            for j in range(min(i, m - 1), max(m - n + i, 0) - 1, -1):
                if x == t[j]:
                    f[j + 1] += f[j]
        return f[m]


class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n = len(s)
        m = len(t)

        dp = [[0] * (m + 1) for _ in range(n + 1)]
        for i in range(n + 1):
            dp[i][0] = 1
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                dp[i][j] = dp[i - 1][j]
                if s[i - 1] == t[j - 1]:
                    dp[i][j] += dp[i - 1][j - 1]
        return dp[n][m]


class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n = len(s)
        m = len(t)

        @cache
        def dfs(i: int, j: int) -> int:
            if j == m:
                return 1
            if i == n:
                return 0
            res = dfs(i + 1, j)
            if s[i] == t[j]:
                res += dfs(i + 1, j + 1)
            return res

        return dfs(0, 0)
